packages/sitecore-jss/src/layout/utils.ts

```diff
@@ -72,3 +79,41 @@ export function getChildPlaceholder(
 
   return rendering.placeholders[placeholderName];
 }
+
+/**
+ * Walks through layout rendering tree and invokes Reduce-compatible callback on every Rendering.
+ * Initial value determines return type
+ *
+ * @param {LayoutServiceData | RouteData<any> | ComponentRendering | HtmlElementRendering | null } object Layout, route data or rendering
+ * @param {Function} callback Reduce-style callback, accepting previous value and a rendering
+ * @param {any} initialValue Initial value that will be passed to callbacks, e.g. array
+ */
+export function reduceLayout<T>(
+  object: LayoutServiceData | RouteData<any> | ComponentRendering | HtmlElementRendering | null,
+  callback: (
+    previousValue: T,
+    component: ComponentRendering | HtmlElementRendering | RouteData
+  ) => T,
+  initialValue: T
+): T {
+  if (object === null) return initialValue;
+
+  // When layout: Check placheolders
+  if ('sitecore' in object) {
+    return reduceLayout(object.sitecore.route, callback, initialValue);
+  }
+
+  // When rendering / HTML Rendering / RouteData, invoke callback
+  initialValue = callback(initialValue, object);
+
+  // When route / component: check its placeholders
+  if ('placeholders' in object) {
+    initialValue = Object.values(object.placeholders || {})
+      .flat()
+      .reduce((previousValue, rendering) => {
+        return reduceLayout(rendering, callback, previousValue);
+      }, initialValue);
+  }
+
+  return initialValue;
```

Comment:
I like the idea of the layout reducer, but still not 100% convinced it's worth the extra complexity with type checking, etc. :) Maybe I need to think about other use cases ...

You could introduce custom type guard functions for a lot of these "in" cases you've got. Might clean things up? We use this approach elsewhere in the repo ([example](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-forms/src/FormField.ts#L12)).

We would also need unit test coverage for this if we went forward with it.

Reply:
I agree with you that it's not the nicest way to do it, but my thinking process went like this:
1) I didn't want to spread this feature across too many files and packages. I'm a guest here. So i did introduce a couple generic utility functions just out of courtesy, as i think reduceLayout can be useful (including for personalization code that i riffed on). I used `getFieldValue` helper though to make code a bit less obnoxious though
2) I dont know what are the conventions for intrducing type guards all around. Typescript does a decent job narrowing the types when you hover mouse over variables inside those conditional branches. 
3) The complexity is inherent to the data structure. By hiding it all in a function, we can reason about it as a black box and it's actually simplier than having many functions. 

So even though I agree with you general premise that i could ber better, i didn't want to make this bigger than adding functions. I'll clean up what I can though

Reply:
Thanks for the detailed thought process. I'm coming around :) 

Totally makes sense for you to keep a light touch as a guest. What I'm seeing in the latest is looking pretty clean. Any type guards or type consolidation we can handle when we take over (I recall some cleanup for duplicate fields type [recently](https://github.com/Sitecore/jss/pull/1297/files#diff-7dc4ee171f0b0a809ec94416824cc71df193bf76a0605c8e8504cf9fc715530e) actually).

Will look for final draft.